src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1572 @@
+            if (isStatic)
+            {
+                // Don't add base class constructors
+                if (filter == null)
+                {
+                    filterToCall = o => !IsConstructor(o);
+                }
+                else
+                {
+                    filterToCall = o => !IsConstructor(o) && filter(o);
+                }
+
+                if (!foundConstructor)
+                {
+                    results.Add(
+                        new CompilerGeneratedMemberFunctionAst(PositionUtilities.EmptyExtent, typename.TypeDefinitionAst,
+                            SpecialMemberFunctionType.DefaultConstructor));
+                }
+            }
+            else
+            {
+                // Reset the filter because the recursive call will add IsHidden back if necessary.
+                filterToCall = filter;
+            }
+            results.AddRange(GetMembersByInferredType(new PSTypeName(typeof(object)), isStatic, filterToCall));
+        }
+
+        private void AddMembersByInferredTypesClrType(PSTypeName typename, bool isStatic, Func<object, bool> filter, Func<object, bool> filterToCall,
+            List<object> results)
+        {
+            if (CurrentTypeDefinitionAst == null || CurrentTypeDefinitionAst.Type != typename.Type)
+            {
+                if (filterToCall == null)
+                {
+                    filterToCall = o => !IsMemberHidden(o);
+                }
+                else
+                {
+                    filterToCall = o => !IsMemberHidden(o) && filter(o);
+                }
+            }
+            IEnumerable<Type> elementTypes;
+            if (typename.Type.IsArray)
+            {
+                elementTypes = new[] {typename.Type.GetElementType()};
+            }
+            else
+            {
+                elementTypes = typename.Type.GetInterfaces().Where(
+                    t => t.GetTypeInfo().IsGenericType && t.GetGenericTypeDefinition() == typeof(IEnumerable<>));
+            }
+            foreach (var type in elementTypes.Prepend(typename.Type))
+            {
+                // Look in the type table first.
+                if (!isStatic)
+                {
+                    var consolidatedString = DotNetAdapter.GetInternedTypeNameHierarchy(type);
+                    results.AddRange(ExecutionContext.TypeTable.GetMembers<PSMemberInfo>(consolidatedString));
+                }
+
+                var members = isStatic
+                    ? PSObject.dotNetStaticAdapter.BaseGetMembers<PSMemberInfo>(type)
+                    : PSObject.dotNetInstanceAdapter.GetPropertiesAndMethods(type, false);
+                results.AddRange(filterToCall != null ? members.Where(filterToCall) : members);
+            }
+        }
+
+        internal IEnumerable<PSTypeName> InferType(Ast ast, TypeInferenceVisitor visitor)
+        {
+            var res = ast.Accept(visitor);
+            if (res == null)
```

Comment:
I don't think it ever will. This is a case where I wish I could express non-nullable reference types in the type system, but since the signature is an object, I check it. Maybe replace with an assert?